### PR TITLE
fix median and percentile aggregators bugs when used as reducer

### PR DESCRIPTION
**Pull request by pi-r-p**
- median aggregator was not usable as reducer, or returned the wrong result.
- position median was done on raw hhcode

example of non working code : 
https://snapshot.senx.io/00059c2130c8d5f8-0-4-5f8666e22df55aa3
https://snapshot.senx.io/00059c21362565e0-0-7-025b60d11920474c

### CODE REVIEW

**Comment by hbs** [comment]
Your modification computes the median for latitude and longitude independently, this is awkward.

**Comment by pi-r-p** [comment]
Indeed... Better than the median of raw HHcodes. Another way would be to compute the median of values, and keep the position of the median point (or centroid in case of even number of points).
Another way is that this mapper/reducer do not return any position.

By the way, reducer.percentile is also buggy.

**Comment by hbs** [comment]
Retaining the location of the actual median point would be more logical, the only issue would be when the median is the mean of two elements and only one of those elements has a location.

File: warp10/src/main/java/io/warp10/script/aggregator/Median.java [hunk 1]
```diff
@@ -26,147 +26,107 @@
...
+      public int compare(Integer idx1, Integer idx2) {
+        if (null == values[idx1] && null == values[idx2]) {
+          return 0;
+        } else if (null == values[idx1] || null == values[idx2]) {
+          return null == values[idx1] ? 1 : -1;
+        } else if (values[idx1] instanceof Number && values[idx2] instanceof Number) {
+          return EQ.compare((Number) values[idx1], (Number) values[idx2]);
         } else {
-          throw new WarpScriptException(this.getName() + " cannot compute median of non numeric values.");
+          throw new RuntimeException("MEDIAN can only operate on numeric Geo Time Series.");
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/aggregator/Median.java]: Retrieve the name from outside the comparator.

File: warp10/src/main/java/io/warp10/script/aggregator/Median.java [hunk 2]
```diff
@@ -26,147 +26,107 @@
...
     //
-    ArrayList<Number> validValues = new ArrayList<Number>(values.length);
-    for (int i = 0; i < values.length; i++) {
-      if (null != values[i]) {
-        if (values[i] instanceof Number) {
-          if (!(values[i] instanceof Double && Double.isNaN((Double) values[i]))) {
-            validValues.add((Number) values[i]);
-          }
+    Arrays.sort(indices, new Comparator<Integer>() {
+      @Override
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/aggregator/Median.java]: Explain strategy for sorting in the presence of nulls

File: warp10/src/main/java/io/warp10/script/aggregator/Median.java [hunk 3]
```diff
@@ -26,147 +26,107 @@
...
+      if (0 == nonNullLength % 2) {
+        //
+        // even number of non null values, return mean of both values.
+        // If there is a location for both points, return centroid of locations
+        // If there is an elevation for both points, return mean of elevations
+        //
+        int low = indices[nonNullLength / 2 - 1];
+        int high = indices[nonNullLength / 2];
+        median = (((Number) values[low]).doubleValue() + ((Number) values[high]).doubleValue()) / 2.0D;
+        if (GeoTimeSerie.NO_ELEVATION != elevations[low] && GeoTimeSerie.NO_ELEVATION != elevations[high]) {
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/aggregator/Median.java]: If there is only one location (same goes for elevation), wouldn't it be better to return that location (resp. elevation)?

File: warp10/src/main/java/io/warp10/script/aggregator/Median.java [hunk 4]
```diff
@@ -26,147 +26,107 @@
...
-    // Could be Long or Double, or a mix, if used as a reducer.
-    //
-    Collections.sort(validValues, new Comparator<Number>() {
-      @Override
-      public int compare(Number o1, Number o2) {
-        return EQ.compare(o1, o2);
-      }
     });
 
+    long location = 0;
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/aggregator/Median.java]: Shouldn't that be GeoTimeSeries.NO_LOCATION?

File: warp10/src/main/java/io/warp10/script/aggregator/Median.java [hunk 5]
```diff
@@ -26,147 +26,107 @@
...
-      int len = validLat.size();
-      if (0 == len % 2) {
-        medianLatitude = (validLat.get(len / 2) + validLat.get(len / 2 - 1)) / 2.0D;
-        medianLongitude = (validLong.get(len / 2) + validLong.get(len / 2 - 1)) / 2.0D;
-      } else {
-        medianLatitude = validLat.get(len / 2);
-        medianLongitude = validLong.get(len / 2);
-      }
-      location = GeoXPLib.toGeoXPPoint(medianLatitude, medianLongitude);
+    if (nullCounter != 0 && this.forbidNulls) {
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/aggregator/Median.java]: 0 != nullCounter

File: warp10/src/main/java/io/warp10/script/aggregator/Median.java (hunk 3 above)
Comment (pi-r-p) [reply warp10/src/main/java/io/warp10/script/aggregator/Median.java]: To be discussed... What is your preference ?

Comment (pi-r-p) [reply warp10/src/main/java/io/warp10/script/aggregator/Median.java]: fixed.

**Comment by pi-r-p** [comment]
- Fixed percentile aggregator with the same method. 
- adds "forbid-nulls" variants to keep the old behaviour.

File: warp10/src/main/java/io/warp10/script/aggregator/Percentile.java [hunk 6]
```diff
@@ -31,150 +31,175 @@
...
       if (0 == n && this.percentile < pn) {
-        return new Object[] { ticks[indices[0]], locations[indices[0]], elevations[indices[0]], values[indices[0]] };
-      } else if (m == indices.length - 1 && this.percentile > pm) {
-        return new Object[] { ticks[indices[m]], locations[indices[m]], elevations[indices[m]], values[indices[m]] };
+        return new Object[]{ticks[indices[0]], locations[indices[0]], elevations[indices[0]], values[indices[0]]};
+      } else if (m == nonNullLength - 1 && this.percentile > pm) {
+        return new Object[]{ticks[indices[m]], locations[indices[m]], elevations[indices[m]], values[indices[m]]};
       } else if (pn == this.percentile) {
-        return new Object[] { ticks[indices[n]], locations[indices[n]], elevations[indices[n]], values[indices[n]] };
+        return new Object[]{ticks[indices[n]], locations[indices[n]], elevations[indices[n]], values[indices[n]]};
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/aggregator/Percentile.java]: Missing space after `Object[]`.

File: warp10/src/main/java/io/warp10/script/aggregator/Percentile.java [hunk 7]
```diff
@@ -31,150 +31,175 @@
...
+      int m = (int) Math.floor(0.5 + this.percentile * nonNullLength / 100.0) - 1;
+
+      double pn = (100.0 / nonNullLength) * (n + 1 - 0.5D);
+      double pm = (100.0 / nonNullLength) * (m + 1 - 0.5D);
 
       if (0 == n && this.percentile < pn) {
-        return new Object[] { ticks[indices[0]], locations[indices[0]], elevations[indices[0]], values[indices[0]] };
-      } else if (m == indices.length - 1 && this.percentile > pm) {
-        return new Object[] { ticks[indices[m]], locations[indices[m]], elevations[indices[m]], values[indices[m]] };
+        return new Object[]{ticks[indices[0]], locations[indices[0]], elevations[indices[0]], values[indices[0]]};
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/aggregator/Percentile.java]: Missing space after `Object[]`.

File: warp10/src/main/java/io/warp10/script/aggregator/Percentile.java [hunk 8]
```diff
@@ -31,150 +31,175 @@
...
+      } else if (m == nonNullLength - 1 && this.percentile > pm) {
+        return new Object[]{ticks[indices[m]], locations[indices[m]], elevations[indices[m]], values[indices[m]]};
       } else if (pn == this.percentile) {
-        return new Object[] { ticks[indices[n]], locations[indices[n]], elevations[indices[n]], values[indices[n]] };
+        return new Object[]{ticks[indices[n]], locations[indices[n]], elevations[indices[n]], values[indices[n]]};
       } else if (pm == this.percentile) {
-        return new Object[] { ticks[indices[m]], locations[indices[m]], elevations[indices[m]], values[indices[m]] };
+        return new Object[]{ticks[indices[m]], locations[indices[m]], elevations[indices[m]], values[indices[m]]};
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/aggregator/Percentile.java]: Missing space after `Object[]`.

File: warp10/src/main/java/io/warp10/script/aggregator/Percentile.java [hunk 9]
```diff
@@ -31,150 +31,175 @@
...
       if (!(value instanceof Number)) {
         throw new WarpScriptException("Invalid parameter for " + getName());
       }
-      
+
       double percentile = ((Number) value).doubleValue();
-      
+
       if (percentile < 0.0D || percentile > 100.0D) {
         throw new WarpScriptException("Invalid percentile, MUST be between 0 and 100.");
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/aggregator/Percentile.java]: Add function name.

**Comment by pi-r-p** [comment]
all fixed

File: warp10/src/main/java/io/warp10/script/aggregator/Percentile.java [hunk 10]
```diff
@@ -67,7 +67,7 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
       double percentile = ((Number) value).doubleValue();
 
       if (percentile < 0.0D || percentile > 100.0D) {
-        throw new WarpScriptException("Invalid percentile, MUST be between 0 and 100.");
+        throw new WarpScriptException("Invalid percentile for" + getName() + ", MUST be between 0 and 100.");
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/aggregator/Percentile.java]: missing whitespace before call to getName

File: warp10/src/main/java/io/warp10/script/aggregator/Median.java [hunk 11]
```diff
@@ -100,7 +100,7 @@ public int compare(Integer idx1, Integer idx2) {
     // singleton case
     //
     if (1 == nonNullLength) {
-      return new Object[]{
+      return new Object[] {
           tick, locations[indices[0]], elevations[indices[0]], values[indices[0]]
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/aggregator/Median.java]: Indentation should be 2 spaces, not 4.

**Comment by pi-r-p** [comment]
all fixed?

**Pull request merged**